[WIP] polys: Introducing FLINT-Backed sparse multivariate polynomials and rings

aasim-maverick opened the pull request:
Via this PR, in `sympy/polys/rings.py`,  

`PolyRing` now delegates to either:

- `FPolyRing` – when the domain is ZZ or QQ and flint is available

- `PyPolyRing`(former `PolyRing`) – otherwise (default pure Python fallback)

As outlined in this #28106, SymPy’s `polys` module currently relies on pure Python implementations for multivariate polynomial rings and their elements. While functional, these implementations struggle with performance in compute-heavy tasks like GCD computation or factorization.

FLINT, a high-performance C library, already provides fast routines for many of these operations, and the python-flint wrapper exposes `fmpz_mpoly` and `fmpq_mpoly` objects that are a great match for our use case.

This PR begins Phase 1 of integrating FLINT-backed multivariate polynomials over `ZZ` and `QQ` into `SymPy`.




#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* polys
  * Added `FPolyRing` and `FPolyElement` classes to enable FLINT-backed sparse polynomial rings and elements over `ZZ` and `QQ`.
  * Modified the `PolyRing` constructor to act as a dispatcher that returns a `FPolyRing` when `flint` is available and the domain is `ZZ` or `QQ`; otherwise falls back to the default `PyPolyRing`.

<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -3094,3 +3446,787 @@ def sqf_list(f, all=False):
+
+    @property
+    def is_irreducible(self):
+        return NotImplementedError
```
NotImplementedError should be raised rather than returned.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -3094,3 +3446,787 @@ def sqf_list(f, all=False):
+        lcm = ground_ring.lcm
+        denom = domain.denom
+
+        for coeff in self.flint_poly.to_dict().values():
+            common = lcm(common, denom(coeff))
```
It is generally bad to convert to dict like this although it may be necessary in some situations if python-flint does not provide the method that we want.

Here we really want fmpq_mpoly to be able to expose the content but I don't know if that is exposed from python-flint yet:
https://flintlib.org/doc/fmpq_mpoly.html#c.fmpq_mpoly_content_ref

The internal representation of the fmpq_mpoly is a single rational number q and an fmpz_mpoly in primitive form. We just need the denominator of `q` in this situation and it should be possible for python-flint to expose that efficiently.

We should keep a list of cases like this and add comments in the code indicating things that could be improved if python-flint provided the ideal interfaces.

aasim-maverick committed:
Setting up superclass-subclass hierarchy for PolyRing

aasim-maverick commented:
Most of the failing tests in the test suite after the setting up of baseclass-subclass hierarchy for `PolyRing`, are for the code breaking in `sympy/polys/ring_series.py` because of the `rs_mul` function which looks something like:

```python
def rs_mul(p1, p2, x, prec):
    R = p1.ring
    p = {}
    if R.__class__ != p2.ring.__class__ or R != p2.ring:
        raise ValueError('p1 and p2 must have the same ring')
    ...
```
The branch comparing `R` and `p2.ring` is entered frequently raising `ValueError` certainly because both of them can be `FlintPolyRing` and `PythonPolyRing` separately making `R.__class__ != p2.ring.__class__` true. I'm not exactly sure why this check is there in the first place because things seem to work fine without it but I'm sure it has some significance since removing this check resolves all the failures but gives way to a new one:

```bash
sympy/polys/tests/test_ring_series.py:141: in test_series_from_list
    r1 = rs_trunc(pc.compose(x, p), x, 5)
sympy/polys/rings.py:1619: in compose
    replacements = [(gens_map[x], ring.ring_new(g)) for x, g in replacements]
sympy/polys/rings.py:1619: in <listcomp>
    replacements = [(gens_map[x], ring.ring_new(g)) for x, g in replacements]
E   KeyError: x
```
The error occurs in `PolyElement.compose` during some generator lookup.  

One last remaining failure that I don't understand is in `sympy/polys/tests/test_factortools.py::test_dup_zz_factor`:
```python
In [1]: R, x = ring("x", ZZ)

In [2]: # The order of the factors will be different when the ground types are
   ...: # flint. At the higher level dup_factor_list will sort the factors.
   ...: c, factors = R.dup_zz_factor(x**3 - 6*x**2 + 11*x - 6)

In [3]: c == 1 # works fine
Out[3]: True

In [4]: set(factors)
Out[4]: {(x - 3, 1), (x - 2, 1), (x - 1, 1)}

In [5]: set(factors) == {(x - 3, 1), (x - 2, 1), (x - 1, 1)} # Should return True but fails
Out[5]: False
```

aasim-maverick commented:
> One last remaining failure that I don't understand is in `sympy/polys/tests/test_factortools.py::test_dup_zz_factor`:
> 
> ```python
> In [1]: R, x = ring("x", ZZ)
> 
> In [2]: # The order of the factors will be different when the ground types are
>    ...: # flint. At the higher level dup_factor_list will sort the factors.
>    ...: c, factors = R.dup_zz_factor(x**3 - 6*x**2 + 11*x - 6)
> 
> In [3]: c == 1 # works fine
> Out[3]: True
> 
> In [4]: set(factors)
> Out[4]: {(x - 3, 1), (x - 2, 1), (x - 1, 1)}
> 
> In [5]: set(factors) == {(x - 3, 1), (x - 2, 1), (x - 1, 1)} # Should return True but fails
> Out[5]: False
> ```

Not sure why this failure occurs but might be some issue with hashing or something else?

aasim-maverick commented:
> The branch comparing `R` and `p2.ring` is entered frequently raising `ValueError` certainly because both of them can be `FlintPolyRing` and `PythonPolyRing` separately making `R.__class__ != p2.ring.__class__` true. I'm not exactly sure why this check is there in the first place because things seem to work fine without it but removing this check resolves all the failures but uncovers another one:
> 
> ```shell
> sympy/polys/tests/test_ring_series.py:141: in test_series_from_list
>     r1 = rs_trunc(pc.compose(x, p), x, 5)
> sympy/polys/rings.py:1619: in compose
>     replacements = [(gens_map[x], ring.ring_new(g)) for x, g in replacements]
> sympy/polys/rings.py:1619: in <listcomp>
>     replacements = [(gens_map[x], ring.ring_new(g)) for x, g in replacements]
> E   KeyError: x
> ```
> 
> The error occurs in `PolyElement.compose` during some generator lookup.

I missed on mentioning it but this has a rather simple fix:
```python
    def compose(self, x, a=None):
        ring = self.ring
        poly = ring.zero

        if a is not None:
            replacements = [(x, a)]
        else:
            if isinstance(x, list):
                replacements = list(x)
            elif isinstance(x, dict):
                replacements = sorted(x.items(), key=lambda k: ring.index(k[0])) # using ring.index instead of gens_map
            else:
                raise ValueError(
                    "expected a generator, value pair a sequence of such pairs"
                )

        replacements = [(ring.index(x), ring.ring_new(g)) for x, g in replacements] # using ring.index here too 

        return self._compose(replacements, initial_poly=poly)
```

aasim-maverick committed:
PolyElement.compose refactored to use PolyRing.index instead of gens_map
PolyRing hashing refactored
PolyRing printing refactored

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -219,49 +231,52 @@ class PolyRing(DefaultPrinting, IPolys):
 
     def _init_monomial_operations(self):
-        # Initialize monomial operations based on number of generators.
+        # onitialize monomial operations based on number of generators
```
Typo

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -643,8 +584,83 @@ def _gens(self):
+            return self.clone(symbols=symbols, domain=self.drop(*gens_to_drop))
+
+
+class PythonPolyRing(PolyRing):
+    """Multivariate distributed polynomial rings."""
+
+    def __new__(cls, symbols, domain, order=lex):
+        # validate inputs
+        symbols = tuple(_parse_symbols(symbols))
+        domain = DomainOpt.preprocess(domain)
+        order = OrderOpt.preprocess(order)
+
+        # initialize instance
+        obj = object.__new__(cls)
+        obj._init_instance(symbols, domain, order)
+        return obj
```
I think it is better not to define `__new__` in the subclass. There can be a single `__new__` method in the superclass that preprocesses the arguments and dispatches to a class method like `._new()` for the appropriate subclass. If we have multiple subclasses then they should not duplicate things like `DomainOpt.preprocess(domain)`.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -643,8 +584,83 @@ def _gens(self):
+        obj._init_instance(symbols, domain, order)
+        return obj
+
+    def _gen_index(self, gen):
+        # get generator index from int or str
+        if isinstance(gen, int):
+            if 0 <= gen < self.ngens:
+                return gen
+            else:
+                raise ValueError(f"invalid generator index: {gen}")
+        else:  # gen is a string
+            try:
+                return self.symbols.index(gen)
+            except ValueError:
+                raise ValueError(f"invalid generator: {gen}")
```
Why has this been moved to the subclass?

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -665,6 +681,55 @@ def _from_dict_ground(self, element, orig_domain=None):
+
+        return obj
+
+    def _gen_index(self, gen):
+        # get generator index from int or str
+        # for now, delegate to Python implementation
+        return self._python_ring._gen_index(gen)
```
I don't see why you have moved the method to the python subclass just so that this has to happen. How would the Flint implementation of this method be any different from the Python one?

aasim-maverick replied:
I was hoping we could use [`fmpz_mpoly_ctx.variable_to_index()`](https://python-flint.readthedocs.io/en/latest/fmpz_mpoly.html#flint.fmpz_mpoly_ctx.variable_to_index) or [`fmpz_mpoly_ctx.variable_to_index()`](https://python-flint.readthedocs.io/en/latest/fmpq_mpoly.html#flint.fmpq_mpoly_ctx.variable_to_index) in case of flint here.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -665,6 +681,55 @@ def _from_dict_ground(self, element, orig_domain=None):
         return poly
 
 
+class FlintPolyRing(PolyRing):
+    """Multivariate distributed polynomial rings."""
+
+    def __new__(cls, symbols, domain, order=lex):
+        # validate inputs
+        symbols = tuple(_parse_symbols(symbols))
+        domain = DomainOpt.preprocess(domain)
+        order = OrderOpt.preprocess(order)
+
+        # initialize instance
+        obj = object.__new__(cls)
+        obj._init_instance(symbols, domain, order)
+
+        # for now, store a reference to the equivalent PythonPolyRing
+        # for fallback operations until flint methods are implemented
+        obj._python_ring = PythonPolyRing(symbols, domain, order)
```
The FlintPolyRing needs to have an actual FLINT context as its internal data rather than just a reference to the Python version.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -219,49 +231,56 @@ class PolyRing(DefaultPrinting, IPolys):
-        obj.domain = domain
-        obj.order = order
+        # delegation to appropriate subclass
+        if flint is not None and _supported_flint_domain(preprocessed_domain):
```
Tests need to be added to make sure that FLINT is only used if the order is supported.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -665,6 +681,76 @@ def _from_dict_ground(self, element, orig_domain=None):
+        """Remove specified generator(s) from the ring."""
+        return self._python_ring.drop(*gens)
+
+    def _ring_equality(self, other):
+        """Check ring equality for ring attribute comparison."""
+        # For now, delegate to Python implementation
+        if isinstance(other, FlintPolyRing):
+            return self.flint_ctx == other.flint_ctx
+        else:
+            return self._python_ring._ring_equality(other)
```
We can't compare the flint contexts because there might be different symbols with the same name.

oscarbenjamin commented on sympy/printing/tests/test_repr.py:
```diff
@@ -257,8 +265,11 @@ def test_FracField():
 
 def test_PolyElement():
     R, x, y = ring("x,y", ZZ)
-    assert srepr(3*x**2*y + 1) == "PolyElement(PolyRing((Symbol('x'), Symbol('y')), ZZ, lex), [((2, 1), 3), ((0, 0), 1)])"
 
+    if GROUND_TYPES != 'flint':
+        assert srepr(3*x**2*y + 1) == "PolyElement(PythonPolyRing((Symbol('x'), Symbol('y')), ZZ, lex), [((2, 1), 3), ((0, 0), 1)])"
+    else:
+        assert srepr(3*x**2*y + 1) == "PolyElement(FlintPolyRing((Symbol('x'), Symbol('y')), ZZ, lex), [((2, 1), 3), ((0, 0), 1)])"
```
It would be reasonable to display FLINT in this representation:
```python
In [3]: R.ring
Out[3]: Polynomial ring in x, y over QQ with lex order
```
This is different because it is not a code representation.

aasim-maverick committed:
PolyRing Further refactored
- Printing updated
- SymPy and flint variable name mapping (try1)
- FlintPolyRing.add_gens refactored

aasim-maverick committed:
Flint name mapping refactored to remove potentially unnecessary dicts
add_gens moved back up to the superclass
PolyRing.__eq__ split reversed

aasim-maverick committed:
Setting up pure python hierarchy of PolyElement(works when flint unavailable)

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -3266,33 +4290,36 @@ def decompose(self):
     def sqf_part(self):
-        return self.ring.dmp_sqf_part(self)
+        return self.to_PythonPolyElement().sqf_part()
 
     def sqf_list(self, all=False):
-        return self.ring.dmp_sqf_list(self, all=all)
+        return self.to_PythonPolyElement().sqf_list(all=all)
 
     def factor_list(self):
-        return self.ring.dmp_factor_list(self)
+        return self.to_PythonPolyElement().factor_list()
```
These are returning the wrong type. They should return FlintPolyElement

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -2887,376 +2510,1777 @@ def cancel(self, g):
 
-        S = [1, c]
+    def _gcd_ZZ(self, other):
+        return heugcd(self, other)
 
-        c = -c
+    def _gcd_QQ(self, g):
+        a, b = self.to_PythonPolyElement(), g.to_PythonPolyElement()
+        h, cff, cfg = a._gcd_QQ(b)
+        return (h.to_FlintPolyElement(self.ring),
+                cff.to_FlintPolyElement(self.ring),
+                cfg.to_FlintPolyElement(self.ring))
```
Thank you, I have been thinking of removing them eventually too because, for example in this case, we can just leverage flint's native `gcd` I think.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -2887,376 +2510,1777 @@ def cancel(self, g):
+        self._flint_poly[monom] = coeff
+        self._hash = None
+
+    def __iter__(self):
+        return self.iterterms() # is this the correct way though? using iter(self._flint_poly) fails poly addition.
```
It should be:
```
In [1]: R = QQ[x,y]

In [2]: p = R(x + y)

In [3]: p
Out[3]: x + y

In [4]: list(p)
Out[4]: [(1, 0), (0, 1)]

In [5]: import flint

In [6]: R = flint.fmpq_mpoly_ctx.get(['x', 'y'])

In [7]: x, y = R.gens()

In [8]: p = x + y

In [9]: list(p)
Out[9]: [(1, 0), (0, 1)]
```
Using iterterms here is not correct:
```
In [1]: R = QQ[x,y]

In [2]: p = R(x + y)

In [3]: list(p.iterterms())
Out[3]: [((1, 0), 1), ((0, 1), 1)]
```

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -2887,376 +2510,1777 @@ def cancel(self, g):
+    @classmethod
+    def _new(cls, ring, init):
+        """Create a new FlintPolyElement efficiently."""
+        obj = object.__new__(cls)
+
+        PolyElement.__init__(obj, ring, init)
+
+        if ring.domain.is_ZZ:
+            obj._flint_poly_cls = flint.fmpz_mpoly
+        elif ring.domain.is_QQ:
+            obj._flint_poly_cls = flint.fmpq_mpoly
+        else:
+            raise NotImplementedError(f"Unsupported domain for FlintPolyElement: {ring.domain}")
+
+        obj._flint_poly = obj._create_flint_poly(init, ring.flint_ctx)
+
+        return obj
+
+    def _create_flint_poly(self, init, flint_ctx):
+        poly = self._flint_poly_cls({}, flint_ctx)
+
+        if isinstance(init, dict):
+            init_dict = init
+        elif isinstance(init, (FlintPolyElement, PythonPolyElement)):
+            init_dict = dict(init.items())
+        elif hasattr(init, 'items'):
+            init_dict = dict(init.items())
+        elif hasattr(init, '__iter__') and not isinstance(init, (str, bytes)):
+            try:
+                init_dict = dict(init)
+            except (TypeError, ValueError):
+                init_dict = {}
+        else:
+            init_dict = {}
+
+        for monom, coeff in init_dict.items():
+            if coeff:
+                poly[monom] = coeff
+
+        return poly
```
This all looks very inefficient. Where is the method to create a new element from just a flint mpoly and a ring?

oscarbenjamin commented:
> The merge conflicts need to be resolved.

I have fixed the merge conflicts that were from gh-28177 so now CI checks will run in the PR.

There are test failures all over the place though:
```console
$ pytest sympy/polys/ -nauto --instafail -rA
...
FAILED sympy/polys/tests/test_rings.py::test_PolyElement__iadd_monom - assert (4 == 4 and 4*x^3 + x^2 + y is x^2 + y)
FAILED sympy/polys/tests/test_rootoftools.py::test_CRootOf___new__ - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_rootoftools.py::test_RootSum_free_symbols - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_rootoftools.py::test_CRootOf_attributes - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_rootoftools.py::test_RootSum_doit - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_rootoftools.py::test_RootSum_evalf - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_rootoftools.py::test_RootSum_independent - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_subresultants_qq_zz.py::test_quo_z - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_rootoftools.py::test_RootSum___new__ - AttributeError: 'PythonPolyElement' object has no attribute 'to_PythonPolyElement'. Did you mean...
FAILED sympy/polys/tests/test_modulargcd.py::test_modgcd_bivariate_integers - TypeError: exponent must be an integer, got 0
================ 109 failed, 2236 passed, 7 xfailed, 4 warnings in 163.49s (0:02:43) ================
```
This is because there are many places here that mix up the PythonPolyElement FlintPolyElement types.

Right now the thing to do is just to fix that and make sure that the correct types are returned by each method.

This is the kind of thing that a type checker would easily identify so we really need these classes to have full type annotations. I added a few in gh-28177 but there are many more that should be added here.

aasim-maverick committed:
refactor(polys): Added a PyPolyElement subclass to PolyElement

aasim-maverick committed:
Merge branch 'master' into flint-rings

aasim-maverick committed:
fix doctests(skipped failing ones as of now)

aasim-maverick committed:
fix: mypy errors (try1)

aasim-maverick committed:
perf(polys): Add FPolyElement class

oscarbenjamin commented:
So there are some failures but they mostly seem to come down to this:
```
    def _term_div(self):
>       raise NotImplementedError
E       NotImplementedError
```
I would have thought that the `_term_div` method should not be needed in the Flint subclass because any other method that would call it should either use FLINT or should convert to the Python representation first.

At this stage it is best to just get to make the minimal changes needed from here to get from here to a working state.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -2694,236 +2490,2211 @@ def terms(self, order=None):
+    def _mul(self, other):
+        python_ring = self._get_python_ring()
+
+        self_python = python_ring.from_flint(self)
+        if isinstance(other, FPolyElement):
+            other_python = python_ring.from_flint(other)
+        else:
+            other_python = other
```
Should it not be guaranteed that other is `FPolyElement`?

Allowing the wrong type here seems like a mistake because it is letting buggy outputs propagate. There is quite a bit of code that is doing this but I think it is backwards. You should be trying to identify all the places where the wrong output is returned rather than modifying other places to expect the wrong output. It is better if this fails with an error so that you can see as early as possible when the wrong type has appeared.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -2694,236 +2490,2211 @@ def terms(self, order=None):
+                else:
+                    raise TypeError(f"Expected PolyElement, got {type(poly)}")
+
+        # Perform remainder computation using PyPolyElement's _rem method
+        quo_py = self_py.quo(G_py)
+        #print("QUO PYTHON", quo_py)
+        # Convert result back to FPolyElement
+        if isinstance(quo_py, list):
+            quo_flint = [self.ring.from_python(q) for q in quo_py]
+        else:
+            quo_flint = self.ring.from_python(quo_py)
```
This kind of confusion about the types like whether this is a list or not means that you have not cleanly divided things between the superclass and the subclass. The method here in the subclass should have well defined input and output types. It should be the superclass that checks whatever is needed to separate the different cases and then call the well-typed methods that are implemented in the subclass.

aasim-maverick committed:
Fixing failures:

- bootstrapped div
- fixed failing tests in test_rings.py
- fixed failures in test_ring_series.py
- skipped some doctests in in docs, rings.py

aasim-maverick committed:
Fixing failures:
- bootstrapped div
- fixed failing tests in test_rings.py
- fixed failures in test_ring_series.py
- skipped some doctests in in docs, rings.py (can discuss further
  if there's a better way to handle that)

aasim-maverick committed:
fixing __mul__ test to run correctly with or without flint

oscarbenjamin commented on sympy/polys/tests/test_rings.py:
```diff
@@ -822,8 +824,13 @@ def test_PolyElement___mul__():
 
     assert dict(3*u*(x + y) + z) == dict((x + y)*3*u + z) == {(1, 0, 0): 3*u, (0, 1, 0): 3*u, (0, 0, 1): 1}
 
-    raises(TypeError, lambda: t*x + z)
-    raises(TypeError, lambda: x*t + z)
+    if GROUND_TYPES != "flint":
+        raises(TypeError, lambda: t * x + z)
+        raises(TypeError, lambda: x * t + z)
+    else:
+        raises(NotImplementedError, lambda: t * x + z)
+        raises(NotImplementedError, lambda: x * t + z)
```
Rather than changing the test like this it would be better to change the code so that it raises the correct exception. Otherwise this is potentially just going to cause more problems down the line where the wrong exception is raised as part of something that is not just a test.

For now though this should just be converting to the Python representation so there should not be any difference in behaviour.

oscarbenjamin replied:
The point is to fix the code so that the test passes. Either catch the NotImplementedError and raise TypeError or use the python implementation rather than the flint implementation.

oscarbenjamin replied:
This is the code that now raises a different exception:
```python
from sympy import ring, ZZ

Rt, t = ring("t", ZZ)
Ruv, u,v = ring("u,v", ZZ)
Rxyz, x,y,z = ring("x,y,z", Ruv)

e = t*x + z
```
This is what happens on the master branch:
```console
$ python t.py 
Traceback (most recent call last):
  File "/stuff/current/active/sympy/t.py", line 7, in <module>
    e = t*x + z
        ~^~
TypeError: unsupported operand type(s) for *: 'PolyElement' and 'PolyElement'
```
This is what happens with the PR:
```console
$ python t.py 
Traceback (most recent call last):
  File "/stuff/current/active/sympy/t.py", line 7, in <module>
    e = t*x + z
        ~^~
  File "/stuff/current/active/sympy/sympy/polys/rings.py", line 1057, in __rmul__
    p2 = ring.ring_new(other)
  File "/stuff/current/active/sympy/sympy/polys/rings.py", line 621, in ring_new
    raise NotImplementedError("conversion")
NotImplementedError: conversion
```
You need to figure out what the difference is that causes it to go down a different codepath.

oscarbenjamin replied:
The difference is that when `__mul__` returns `NotImplemented` if both operands are of the same class then `__rmul__` does not get called. In this case though `t` and `x` are of different classes because one is PPolyElement and the other is FPolyElement.

oscarbenjamin replied:
The `__mul__` and `__rmul__` methods will need to be cleaned up. We can discuss this in the meeting later.

aasim-maverick committed:
reverting __mul__ test to original(fix remaining)

oscarbenjamin committed:
fix(polys): Fix PolyElement.__rmul__ to use domain_new

aasim-maverick committed:
refactor(polys): Improving splits(parent-helper) in div and rem; Moved quo and exquo back up in the baseclass

oscarbenjamin commented:
```diff
diff --git a/sympy/polys/ring_series.py b/sympy/polys/ring_series.py
index 4c839eb0e2..64822d2ffd 100644
--- a/sympy/polys/ring_series.py
+++ b/sympy/polys/ring_series.py
@@ -1891,10 +1891,10 @@ def rs_hadamard_exp(p1, inverse=False):
     p = R.zero
     if not inverse:
         for exp1, v1 in p1.items():
-            p[exp1] = v1/int(ifac(exp1[0]))
+            p[exp1] = v1/int(ifac(int(exp1[0])))
     else:
         for exp1, v1 in p1.items():
-            p[exp1] = v1*int(ifac(exp1[0]))
+            p[exp1] = v1*int(ifac(int(exp1[0])))
     return p

 def rs_compose_add(p1, p2):
```

aasim-maverick committed:
fix(polys): converting exponent(exp1[0]) to int in rs_hadamard_exp
